platform/fabric/core/generic/msp/x509/loader.go

```diff
@@ -51,33 +55,99 @@ func GetSigningIdentity(mspConfigPath, mspID string) (SigningIdentity, error) {
 
 // LoadLocalMSPAt loads an MSP whose configuration is stored at 'dir', and whose
 // id and type are the passed as arguments.
-func LoadLocalMSPAt(dir, id, mspType string) (msp.MSP, error) {
+func LoadLocalMSPAt(dir, id, mspType string, bccspConfig *config.BCCSP) (msp.MSP, error) {
 	if mspType != "bccsp" {
 		return nil, errors.Errorf("invalid msp type, expected 'bccsp', got %s", mspType)
 	}
 	conf, err := msp.GetLocalMspConfig(dir, nil, id)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "could not get msp config from dir [%s]", dir)
 	}
-	ks, err := sw.NewFileBasedKeyStore(nil, filepath.Join(dir, "keystore"), true)
+
+	cp, _, err := GetBCCSPFromConf(dir, bccspConfig)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "failed to get bccsp from config [%v]", bccspConfig)
 	}
-	cryptoProvider, err := sw.NewDefaultSecurityLevelWithKeystore(sw.NewDummyKeyStore())
-	if err != nil {
-		return nil, err
+
+	mspOpts := &msp.BCCSPNewOpts{
+		NewBaseOpts: msp.NewBaseOpts{
+			Version: msp.MSPv1_0,
+		},
 	}
-	thisMSP, err := msp.NewBccspMspWithKeyStore(msp.MSPv1_0, ks, cryptoProvider)
+	thisMSP, err := msp.New(mspOpts, cp)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "failed to create new BCCSPMSP instance at [%s]", dir)
 	}
 	err = thisMSP.Setup(conf)
 	if err != nil {
-		return nil, err
+		return nil, errors.WithMessagef(err, "failed to setup the new BCCSPMSP instance at [%s]", dir)
 	}
 	return thisMSP, nil
 }
 
+// GetBCCSPFromConf returns a BCCSP instance and its relative key store from the passed configuration.
+func GetBCCSPFromConf(dir string, conf *config.BCCSP) (bccsp.BCCSP, bccsp.KeyStore, error) {
+	if conf == nil {
+		return GetSWBCCSP(dir)
+	}
+
+	switch conf.Default {
+	case "SW":
+		return GetSWBCCSP(dir)
+	case "PKCS11":
+		return GetPKCS11BCCSP(conf)
+	default:
+		return nil, nil, errors.Errorf("invalid config.BCCSP.Default.%s", conf.Default)
+	}
+}
+
+// GetPKCS11BCCSP returns a new instance of the HSM-based BCCSP
+func GetPKCS11BCCSP(conf *config.BCCSP) (bccsp.BCCSP, bccsp.KeyStore, error) {
+	if conf.PKCS11 == nil {
+		return nil, nil, errors.New("invalid config.BCCSP.PKCS11. missing configuration")
+	}
+
+	p11Opts := *conf.PKCS11
+	ks := sw.NewDummyKeyStore()
```

Comment:
why are we using the DummyKeyStore from the bccsp sw package?

Reply:
Is this really the correct way of using the `pkcs11.New` method?

Reply:
this is the way it is instantiated in Fabric. The PKCS11 BCCSP has internally a SW BCCSP that needs a key store. In Fabric, they pass a dummy key store. I assumed it is okay for us as well.